Use new codecov uploader

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Codecov deprecated their bash uploader for reasons in favor of a new Javascript uploader. It doesn't support all the flags the bash uploader did.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Test Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`.azure-pipelines/scripts/publish-codecov.sh`

samdoran committed:
Use new codecov uploader

Download the new uploader from our S3.
Remove flags that are invalid in the new uploader.

ci_complete
ci_coverage

samdoran committed:
Set slug in upload command

This should be detected automatically but that doesn't seem to be the case.

ci_complete
ci_coverage

samdoran committed:
Setting the slug didn't help

This appears to be a bug in the new uploader on the Codecov API parsing data
from Azure Pipelines.

webknjaz commented:
Watch out for the sudden bugs, though: https://github.com/codecov/uploader/pull/413. The new uploader repo clearly states that it's WIP in their README.

samdoran committed:
Codecov fixed the API 🙌

ci_complete
ci_coverage

samdoran committed:
Add a slug to the codecov command

ci_complete
ci_coverage

samdoran commented:
https://github.com/codecov/uploader/pull/426/files fixes this in the new uploader itself.

samdoran committed:
Do not add slug

The need to specify the slug in Azure Pipelines was fixed in version 0.1.6
of the new uploader.

ci_complete
ci_coverage

samdoran committed:
Rewrite coverage uploader in Python

ci_complete
ci_coverage

samdoran committed:
Group coverage by command and upload with flags

samdoran committed:
Fix sanity test failures

ci_complete
ci_coverage

samdoran committed:
The flags parameter is not a comma separated list

samdoran committed:
The path containing coverage files is already a Path object

samdoran committed:
Use temporary file for codecov binary

ci_coverage
ci_complete

samdoran committed:
Properly tokenize flags parameter

samdoran committed:
Make flags parameter required and flush buffer

ci_coverage
ci_complete

samdoran committed:
Pass TemporaryFile object into function rather than Path object

Pass TemporaryFile object into function rather than Path object

samdoran committed:
Change comment to docstring

samdoran committed:
Cleanup imports

Import stanard lib items directly
Put alias import in with other imports.

samdoran committed:
Use a temporary directory rather than a temproray file

Certain file systems do not allow files to be executed while they are open for
writing. Closing a NamedTemporaryFile by default will remove the file, which
means it would not be available for execution. Use a TemporaryDirectory instead.

samdoran committed:
Make list comprehension a bit easier to read

ci_complete
ci_coverage

samdoran committed:
No need to flush an already closed file

samdoran committed:
Preserve coverage file grouping

ci_coverage
ci_complete
